File: lib/src/replay/replay_runtime.dart
import 'dart:convert';

import '../graph/runtime_graph.dart';
import 'replay_equivalence.dart';
import 'replay_fingerprint.dart';
import 'replay_graph.dart';
import 'replay_memory.dart';

Map<String, dynamic> replayRuntimeState(Map<String, dynamic> extraction) =>
    jsonDecode(jsonEncode(extraction)) as Map<String, dynamic>;

Map<String, dynamic> validateFullRuntimeReplay(
  Map<String, dynamic> original,
  Map<String, dynamic> replayed,
) {
  final graph = _graphFromEnvelope(original);
  final replayGraph = _graphFromEnvelope(replayed);
  final replay = validateReplayEquivalence(original, replayed);
  final graphResult = validateGraphReplayEquivalence(graph, replayGraph);
  final mem = original['runtime_memory'] as Map<String, dynamic>?;
  final replayMem = replayed['runtime_memory'] as Map<String, dynamic>?;
  final memory = (mem != null && replayMem != null)
      ? validateMemoryReplayEquivalence(mem, replayMem)
      : null;
  final fingerprint =
      validateFingerprintReplayEquivalence(original, replayed, graph: graph, memory: mem);
  final equivalent = replay['equivalent'] == true &&
      graphResult['equivalent'] == true &&
      fingerprint['equivalent'] == true &&
      (memory == null || memory['equivalent'] == true);
  return {
    'equivalent': equivalent,
    'replay': replay,
    'graph': graphResult,
    'memory': memory,
    'fingerprint': fingerprint,
    'bounded': true,
  };
}

RuntimeGraph _graphFromEnvelope(Map<String, dynamic> envelope) {
  final g = envelope['unified_runtime_graph'] ?? envelope['graph'];
  if (g is RuntimeGraph) return g;
  if (g is Map<String, dynamic>) {
    return buildRuntimeGraph({'graph': g});
  }
  return RuntimeGraph(nodes: [], edges: []);
}

File: src/replay/replayRuntime.ts
import type { ExtractionEnvelope } from "../contracts/runtimeContracts.js";
import type { RuntimeGraph } from "../contracts/graphContracts.js";
import { validateReplayEquivalence } from "./replayEquivalence.js";
import { validateGraphReplayEquivalence } from "./replayGraph.js";
import { validateMemoryReplayEquivalence } from "./replayMemory.js";
import { validateFingerprintReplayEquivalence } from "./replayFingerprint.js";

export function replayRuntimeState(extraction: ExtractionEnvelope): ExtractionEnvelope {
  return JSON.parse(JSON.stringify(extraction)) as ExtractionEnvelope;
}

export function validateFullRuntimeReplay(
  original: ExtractionEnvelope,
  replayed: ExtractionEnvelope,
): {
  equivalent: boolean;
  replay: ReturnType<typeof validateReplayEquivalence>;
  graph: ReturnType<typeof validateGraphReplayEquivalence>;
  memory: ReturnType<typeof validateMemoryReplayEquivalence> | null;
  fingerprint: ReturnType<typeof validateFingerprintReplayEquivalence>;
  bounded: boolean;
} {
  const graph = (original.unified_runtime_graph ?? { nodes: [], edges: [] }) as RuntimeGraph;
  const replayGraph = (replayed.unified_runtime_graph ?? replayed.graph ?? graph) as RuntimeGraph;
  const mem = (original as Record<string, unknown>).runtime_memory as Record<string, unknown> | undefined;
  const replayMem = (replayed as Record<string, unknown>).runtime_memory as
    | Record<string, unknown>
    | undefined;

  const replay = validateReplayEquivalence(original, replayed);
  const graphResult = validateGraphReplayEquivalence(graph, replayGraph);
  const memory =
    mem && replayMem ? validateMemoryReplayEquivalence(mem, replayMem) : null;
  const fingerprint = validateFingerprintReplayEquivalence(original, replayed, graph, mem);

  const equivalent =
    replay.equivalent &&
    graphResult.equivalent &&
    fingerprint.equivalent &&
    (memory === null || memory.equivalent);

  return { equivalent, replay, graph: graphResult, memory, fingerprint, bounded: true };
}
